Client code for ServiceRegistry with ability to set service URI in CLI

yondonfu opened the pull request:
Adds support for storing a service URI as a transcoder/orchestrator in the ServiceRegistry contract.

Dependent on the smart contract changes here: https://github.com/livepeer/protocol/pull/223

Branched off of the current WIP V2 Networking branch because these changes only make sense in the context of the V2 networking architecture.

yondonfu committed:
Client code for ServiceRegistry. Add ability to set service URI in CLI

ericxtang commented on cmd/livepeer_cli/wizard_transcoder.go:
```diff
@@ -30,15 +30,18 @@ func (w *wizard) promptTranscoderConfig() (float64, float64, *big.Int) {
 
-	return blockRewardCut, feeShare, pricePerSegment
+	fmt.Printf("Enter service URI - ")
+	serviceURI := w.readString()
```
Should we validate the URI here to make sure it's actually usable?  Otherwise a typo here can be expensive.

Also, should we enforce https?  Seems like a good thing, but adds some operational complexity.  @j0sh do you have https thoughts since you've been working on the new networking scheme?

j0sh replied:
Setting the Service URI is probably going to require a good bit of education and documentation for new and existing transcoders, so they understand what this entails. We could auto-select the value [0] and have the user confirm, to make it easier all-around. Wasn't expecting that feature in this basic PR though.

About HTTPS, it's actually easier to get a combined grpc+plain-http handler on the same port if TLS is enabled. We already generate self-signed certs at runtime [1], and set the client to ignore them [2]. Once we settle on an approach for https://github.com/livepeer/go-livepeer/issues/437 then we can start checking the certs rather than ignoring them.

[0] By finding the transcoder's own public address, then prefixing the default scheme. Also, print a loud warning if the transcoder appears to be NAT'd.
[1] https://github.com/livepeer/go-livepeer/pull/456/commits/5627be7d537cc46bf3016f01789cc374d9debb26
[2] https://github.com/livepeer/go-livepeer/pull/447/commits/066dcf7e081657958c0966f1245bcfe61c222689#diff-44bb03d888e11d6429223a7b0e55220dR204

yondonfu replied:
Agreed that there should definitely be validation for the service URI since the contract will store arbitrary strings. For starters, we could probably use [url.ParseRequestURI](https://golang.org/pkg/net/url/#ParseRequestURI) just to check if the URI is valid

j0sh commented:
Thanks for this -- I've cherry-picked the commit into https://github.com/livepeer/go-livepeer/pull/447 . While this PR looks good from my end, I'll fixup my branch as needed based on feedback to this PR.

yondonfu committed:
Check for valid service URI input from user

yondonfu commented:
Hm looks like the commit history for this PR got a little messy on the most recent push...This is the relevant newest commit which contains some basic URI validation on user input: https://github.com/livepeer/go-livepeer/pull/456/commits/51e6449a7a30a285a4c3775ec05a5e2bac845d96

j0sh commented:
Cherry-picked https://github.com/livepeer/go-livepeer/commit/51e6449a7a30a285a4c3775ec05a5e2bac845d96 into my branch, thanks

j0sh committed:
cli: Set a default ServiceURI.

Call out to a IP address service to find the node's public IP.

Also add https to the input, since that is the only scheme for now.

j0sh committed:
livepeer: remove publicIP option.

No longer needed with new networking protocol.

j0sh committed:
livepeer: Verify ServiceURI matches on startup.

Also add an option to bypass IP checking by passing in -publicAddr.

Enforcement is disabled for now while transcoders set the URI.

j0sh commented:
Added a few more commits around the ServiceURI. I'd like to get consensus on these changes since it's something that affects how users deal with command line arguments (ergo, a breaking change if anyone has scripts to automate node setup). Happy to spin this off into a separate PR if that works better.

Two things prompted these changes:
* Usability concerns such as https://github.com/livepeer/go-livepeer/issues/449 
* In the new networking protocol, we don't have a fallback method of talking to the transcoder through the p2p network. So we need to ensure that the transcoder's public information is up-to-date.

https://github.com/livepeer/go-livepeer/pull/456/commits/6d637af8c6e3da0c6a1c6402c3679cb6e85742bb : Usability improvement in auto-discovering the public IP and informing the user that their node needs to be rebooted in order to pick up changes to ServiceURI. Note that the `https://` scheme is not required from the user here anymore. A hostname is also acceptable.

https://github.com/livepeer/go-livepeer/pull/456/commits/f3d0bec3250d3994011a249fadac6e79737368b4 : `publicIP` not needed anymore.

https://github.com/livepeer/go-livepeer/pull/456/commits/483571188a904a78364503056101326810aa8d19 : Checks the ServiceURI matches the public IP on startup. Ideally we'd kill the node if it doesn't match (since the transcoder would then be inaccessible), but we need to give active transcoders time to migrate, so we only print a warning for now.

Also offers a `publicAddress` option to bypass the IP check, but the publicAddress still needs to match what's in ServiceURI. The reason we aren't reusing the `publicIP` option here is because this could be a hostname, and is also useful in other contexts, eg devenv or private networks.

We really need to communicate to our transcoders (through release notes, Discord, forums, readme,  readthedocs, wiki) about the new public address requirement and the need to set the Service URI. I'll write up something so folks aren't caught by surprise once the release hits, but this set of commits needs to be signed off first.

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -191,15 +193,7 @@ func main() {
-		return
-	}
-	nw, err := bnet.NewBasicVideoNetwork(node, *publicIP, *port)
+	nw, err := bnet.NewBasicVideoNetwork(node, "127.0.0.1", *port)
```
Do we default to passing in 127.0.0.1 with the expectation that with the V2 networking changes, this basicnet construct won't be used anyway so it doesn't matter that a public IP isn't set for this?

j0sh replied:
Correct, the basicnet publicIP won't be used anymore with the v2 network. Technically this API could be cleaned up too, but I'm not sure how much further work the p2p layer will see; it depends what we do with the relay network (if anything).

yondonfu commented on cmd/livepeer_cli/wizard_transcoder.go:
```diff
@@ -64,24 +94,30 @@ func (w *wizard) activateTranscoder() {
 
 	httpPostWithParams(fmt.Sprintf("http://%v:%v/setTranscoderConfig", w.host, w.httpPort), val)
+	// TODO we should confirm if the transaction was actually sent
+	fmt.Println("\nTransaction sent. Once confirmed, please restart your node if the ServiceURI has been reset")
```
The reset is just to allow the node to check for you if your service URI matches your publicAddr before moving forward with any other operations?

j0sh replied:
The restart also allows the node to bind to the new port, if that has changed.

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -506,6 +500,35 @@ func setupTranscoder(ctx context.Context, n *core.LivepeerNode, em eth.EventMoni
+	if uri.Hostname() != publicAddr {
+		glog.Errorf("Service address %v did not match discovered address %v; set the correct address in livepeer_cli or use -publicAddr", uri.Hostname(), publicAddr)
+		// TODO remove '&& false' after all transcoders have set a service URI
+		if active && false {
```
Hm so after we remove the `&& false` to kill the node if an active transcoder has mismatched service URI + publicAddr if it is still the case that an active transcoder did not properly set a serviceURI to match their publicAddr they won't be able to use the CLI to set a service URI anymore since the transcoder configuration options would only be available when the node is running with the `-transcoder` flag, but if the account for the node is already an active transcoder without a service URI then every time the node boots up with the `-transcoder` flag the node will be killed. I guess the workaround here is that the user could pass in `-publicAddr 127.0.0.1` so that it matches the default service URI, successfully boot up the node, set an actual service URI and the reboot the node without the `-publicAddr` flag.

Not really a big concern IMO since the CLI makes you pick a service URI upon transcoder registration (the scenario would really only happen if someone registered as a transcoder outside of the CLI and chose to leave out the service URI), but just thought I would note the scenario.

j0sh replied:
Correct, the `&&false` is to give currently active transcoders an opportunity to set up their service URIs; new transcoders registering after the network switchover wouldn't have this issue.

Also correct that `-publicAddr 127.0.0.1`  would be the fallback if current transcoders still don't set a service URI. Note that transcoders will still start up by default on port 4433 if there is no service URI set; see https://github.com/livepeer/go-livepeer/pull/447/files#diff-236217e7f473dfbde9134c6a391f3408R126

Since we have the fallback, we could still kill the node by default to force transcoders to make the adjustment, but I figured it's nicer to be a little less abrupt during the transition. We probably should monitor this though. The current transcoder pool is small enough that we could reach out to individual transcoders after a few weeks, if needed.

yondonfu reviewed (commented):
Changes make sense to me - I agree that communication around the new service URI requirement will be really important
